sympy/utilities/timeutils.py

```diff
@@ -74,3 +82,143 @@ def wrapper(*args, **kwargs):
+    ...     res = solve(expr, x)
+    ...
+    TimeoutError: Execution time exceeded 2 s.
+
+    >>>
+    >>> solutions = TimeoutThreading.map(solve, exprs, dep_vars, seconds_per_item=2)
+    """
+
+    def __init__(self, duration_seconds: int):
+        self.duration_seconds = duration_seconds
+
+    @classmethod
+    def exception_type(cls) -> type[Exception]:
+        return TimeoutError
+
+    @classmethod
+    def map(cls, func: Callable[..., ...], *args, seconds_per_item: int):
+        for tup in zip(*args):
+            try:
+                with cls(duration_seconds=seconds_per_item):
+                    yield func(*tup)
+            except cls.exception_type() as e:
+                yield e
+
+    @property
+    def elapsed(self) -> float:
+        """Elapsed time in seconds."""
+        return time.time() - self._t0
+
+    @property
+    def remaining(self) -> float:
+        """Remaining time in seconds."""
+        return self.duration_seconds - self.elapsed
+
+
+    def exception_message(self) -> str:
+        return f"Execution time exceeded {self.duration_seconds} s."
+
+    def _construct_exception(self) -> Exception:
+        return self.exception_type()(self.exception_message)
+
+    def __enter__(self):
+        self._t0 = time.time()
+
+
+class TimeoutThreading(_TimeoutContextManager):
+    def __enter__(self):
+        super().__enter__()
+        self._queue = queue.Queue()
+        self.timer_thread = threading.Timer(self.duration_seconds, self._terminate)
+        self.timer_thread.start()
+        return self
+
+    def _terminate(self):
+        self._queue.put(self._construct_exception())
+        _thread.interrupt_main()  # raises KeyboardInterrupt on the main thread.
+
+    def __exit__(self, *exc):
+        try:
+            exc = self._queue.get(block=False)
+        except queue.Empty:
+            self.timer_thread.cancel()
+            return False
+        else:
+            # there will be spurious KeyboardInterrupt exception in the traceback
+            # if one really wanted to, it could be edited away, e.g.:
+            # https://github.com/pallets/jinja/blob/9dad679695ccf84f06f8df897a6c5908e1963363/src/jinja2/debug.py#L14
+            raise exc
+
+
+class TimeoutSignal(_TimeoutContextManager):
+    """Uses an ALARM signal (for POSIX compliant systems) to limit execution time."""
+    def __enter__(self):
+        super().__enter__()
+        if os.name == 'nt':
+            warnings.warn("Windows might not support ALARM signal, prefer TimeoutThreading instead.")
+        signal.signal(signal.SIGALRM, self._terminate)
+        signal.alarm(self.duration_seconds)
+        return self
+
+    def _terminate(self, signum, frame):
+        assert signum == signal.SIGALRM
+        raise self._construct_exception()
+
+    def __exit__(self, *exc):
+        signal.alarm(0)
+        return False
+
+
+default_timeout_context_manager_class = TimeoutThreading if os.name == 'nt' else TimeoutSignal
```

Comment:
Later I noticed `hasattr(signal, 'SIGALRM') == False` on windows, that is probably the more robust solution in case there are other non-posix OS'es we support?